Changed WASD to keybindings

stinky-lizard opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Removes hardcoded WASD keys and replaces them with WASD keybinds. Now, instead of just using WASD, you can use any key on the keyboard - they'll even fire alongside other keybinds. Uses the existing keybind system. WASD is still default.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Allows people with non-QWERTY keyboard layouts to play much more comfortably. In turn, fixes #3633 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Made WASD movement keys rebindable
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

stinky-lizard committed:
Changed WASD to keybindings

Kapu1178 commented:
Does this maintain non-hotkey mode movement?

stinky-lizard commented:
By the way, did a quick test on a local server, and worked smoothly. Might not hold up to lag though? Rebinding WASD to -P[] worked fine.

stinky-lizard commented:
> 
> 
> Does this maintain non-hotkey mode movement?

If you mean the arrow keys, yes, it does.

stinky-lizard commented:
Here's a quick video showcasing it: https://youtu.be/8NYEZZGI8NA

francinum commented:
Please use the *exact phrasing* of `closes #000` or `fixes #000`, with each individual issue having it's own `closes/fixes/resolves`
Github doesn't track closure properly if you don't do this.

stinky-lizard commented:
Oh, sorry. Edited.

stinky-lizard committed:
right forgot about updating the savefile

francinum commented:
Looks good overall.

KubeRoot commented on code/controllers/subsystem/input.dm:
```diff
@@ -95,7 +95,7 @@ SUBSYSTEM_DEF(input)
 /datum/controller/subsystem/input/proc/setup_default_movement_keys()
 	var/static/list/default_movement_keys = list(
-		"W" = NORTH, "A" = WEST, "S" = SOUTH, "D" = EAST,				// WASD
+//		"W" = NORTH, "A" = WEST, "S" = SOUTH, "D" = EAST,				// WASD are now keybinds, not hard-coded.
```
The only thing I'm unsure about is if this should stay commented out, or just be removed.

stinky-lizard replied:
That's fair. I've removed it.

KubeRoot reviewed (dismissed):
Looks good, tested with both hotkey and old-style input.

stinky-lizard committed:
Removed superfluous comment

KubeRoot reviewed (changes requested):
This PR needs to be updated so the new keybindings match how keybindings are coded now: https://github.com/BeeStation/BeeStation-Hornet/pull/5672

This generally means each keybinding no longer does its own checks `if(!user.mob) return`, but calls parent and checks the return value of that.

I also didn't notice earlier, but it seems keybindings should return `TRUE` if the input was handled correctly, which none of the new keybindings do; is there a specific reason for that?

stinky-lizard committed:
Merge branch 'BeeStation:master' into master

stinky-lizard committed:
updated to match with new keybinding system

the new keybinding system didn't change preferences_savefile.dm or input.dm, so i left them alone. i updated the new keybinds in mob.dm to match the others and added new signals to keybinding.dm.

although, it seems mob.dm is a bit... inconsistent, as prevent_movement doesn't return true, and prevent_movement/up checks its parent when toggle_move_intent/up doesn't. I decided to add both the check and the return to the new /up's, but im not sure if that's the best
tbh, im mostly copying the other keybinds ¯\_(ツ)_/¯

stinky-lizard commented:
> This PR needs to be updated so the new keybindings match how keybindings are coded now: #5672
> 
> This generally means each keybinding no longer does its own checks `if(!user.mob) return`, but calls parent and checks the return value of that.
> 
> I also didn't notice earlier, but it seems keybindings should return `TRUE` if the input was handled correctly, which none of the new keybindings do; is there a specific reason for that?

I've updated the new keybinds to match how the others act in mob.dm - neither of the other two changed files in this commit  (input.dm & preferences_savefile.dm) were touched by #5672, so I figured everything in them would be communicated with in the same way, so I think they should be fine without changes?

Also, all the keybindings still have `if(!user.mob) return`... I checked and that's not just from my commits, they're there in the main fork. So I kept those in my keybinds as well.

There were some other oddities as well, like how prevent_movement didn't return `TRUE` at all, or how `prevent_movement/up` checks its parent but `toggle_move_intent/up` doesn't. Is that intentional? I barely have a grasp on how the whole system works, I'm mostly copying the other keybinds, but y'know, can you blame me?

Either way, tested in a local server and it worked fine. ¯\_(ツ)_/¯

stinky-lizard commented:
Oh, also, I noticed this fixes a bit of a movement bug: without this PR, moving with WASD and switching to non-hotkey mode while moving keeps you moving in that direction until you return to hotkey mode and press the button again - pressing the arrow keys wont fix it. With it, switching from hotkey mode still keeps you moving, but now you only have to press the corresponding direction on the arrow keys.

KubeRoot commented:
> Also, all the keybindings still have `if(!user.mob) return`... I checked and that's not just from my commits, they're there in the main fork. So I kept those in my keybinds as well.

Uh oh, certainly not all the keybindings, and I believe that's my fault, gonna have to look into it. If I remember correctly, these should only be there for very peculiar keybindings that require checks that don't fit in their type. The reason for this, `keybinding/can_use` is ran to decide if `/down` or `/up` is actually executed, and those are overriden on subtypes like `keybinding/mob` to include the checks previously duplicated in each `down`/`up`.

stinky-lizard commented:
> Uh oh, certainly not all the keybindings, and I believe that's my fault, gonna have to look into it. If I remember correctly, these should only be there for very peculiar keybindings that require checks that don't fit in their type.

Upon closer inspection, not all of them have it, but most do, at least in `mob.dm`.

utilbird commented on code/datums/keybinding/mob.dm:
```diff
@@ -1,8 +1,104 @@
+	. = ..()
+	if(.)
+		return
+	if(!user.mob) return
```
Properly newline indent the returns for consistency with the rest of the codebase

stinky-lizard replied:
On line 17? The rest of the file has the same indentation as 14-17, should I change those as well?

PowerfulBacon reviewed (approved):
Works fine
